### PR TITLE
[vnetorch] Add ECMP support for vnet tunnel routes

### PR DESCRIPTION
**Pull request by shi-su**
**What I did**
Add functions to create/remove next hop groups for vnet tunnel routes.
Count the reference count of next hop groups to create and remove as needed.
Share the counter of next hop groups with routeorch.
Add vs test

**Why I did it**
To add support for overlay ECMP.

**How I verified it**
Verify ECMP groups are properly created and removed with the functions.
Verify vs test passes

**Details if related**

### CODE REVIEW

File: tests/test_vnet.py [hunk 1]
```diff
@@ -1125,6 +1237,118 @@ def test_vnet_vxlan_multi_map(self, dvs, testlog):
...
+
+        # Set the tunnel route to another nexthop group
+        set_vnet_routes(dvs, "fd:8:10::32/128", 'Vnet8', 'fd:8:1::1,fd:8:1::2,fd:8:1::3,fd:8:1::4')
+        route1, nhg1_2 = vnet_obj.check_vnet_ecmp_routes(dvs, 'Vnet8', ['fd:8:1::1', 'fd:8:1::2', 'fd:8:1::3', 'fd:8:1::4'], tunnel_name, route_ids=route1)
+
+        # Check the previous nexthop group is removed
+        vnet_obj.fetch_exist_entries(dvs)
+        assert nhg1_1 not in vnet_obj.nhgs
+
+        # Create another tunnel route to the same set of endpoints
```
Comment (prsunny) [inline tests/test_vnet.py:1327]: add one case for IPv4 prefix over IPv6 endpoints in this same test 8

File: orchagent/vnetorch.cpp [hunk 2]
```diff
@@ -1066,18 +1346,48 @@ bool VNetRouteOrch::handleTunnel(const Request& request)
         }
     }
 
+    if (!vni_list.empty() && vni_list.size() != ip_list.size())
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp:1349]: I thought these conditions are required. The field should be either empty or the match in length. Otherwise, it would be impossible to match the fields with endpoints. For example, if the endpoint had three addresses, the vni field should be one of the choices: empty; `0,0,0`(three valid values); `0,,`(comma separated empty values). The field should not be something like `0,1` since it would be impossible to figure which endpoint it corresponds to.

I also noticed that the current implementation of request parser cannot handle the case of comma-separated empty values. Therefore, I updated the value type of mac and vni such to string so that we can tokenize the values and handle them in vnetorch.

File: orchagent/vnetorch.cpp [hunk 3]
```diff
@@ -1041,23 +1321,23 @@ bool VNetRouteOrch::handleTunnel(const Request& request)
 {
     SWSS_LOG_ENTER();
 
-    IpAddress ip;
-    MacAddress mac;
-    uint32_t vni = 0;
+    vector<IpAddress> ip_list;
+    vector<MacAddress> mac_list;
+    vector<uint64_t> vni_list;
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp]: Updated the vni field type and made it to uint32.

File: orchagent/vnetorch.cpp [hunk 4]
```diff
@@ -648,29 +831,126 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
...
+        {
+            syncd_tunnel_routes_.erase(vnet);
+        }
+
+        vrf_obj->removeRoute(ipPrefix);
     }
     else
     {
-        vrf_obj->removeRoute(ipPrefix);
+        SWSS_LOG_ERROR("Unknown operation");
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp]: Agree, removed this part.

File: orchagent/vnetorch.cpp [hunk 5]
```diff
@@ -648,29 +831,126 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
...
     sai_ip_prefix_t pfx;
     copy(pfx, ipPrefix);
-    sai_object_id_t nh_id = (op == SET_COMMAND)?vrf_obj->getTunnelNextHop(endp):SAI_NULL_OBJECT_ID;
 
-    for (auto vr_id : vr_set)
+    if (op == SET_COMMAND)
     {
-        if (op == SET_COMMAND && !add_route(vr_id, pfx, nh_id))
+        sai_object_id_t nh_id;
+        /* The route in pointing to one single endpoint */
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp]: Thanks for pointing it out. Fixed.

File: orchagent/vnetorch.cpp [hunk 6]
```diff
@@ -648,29 +831,126 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
...
+            if (nexthops.getSize() == 1)
+            {
+                NextHopKey nexthop(nexthops.to_string(), true);
+                NextHopGroupInfo next_hop_group_entry;
+                next_hop_group_entry.next_hop_group_id = vrf_obj->getTunnelNextHop(nexthop);
+                next_hop_group_entry.ref_count = 0;
+                next_hop_group_entry.active_members[nexthop] = SAI_NULL_OBJECT_ID;
+                syncd_nexthop_groups_[vnet][nexthops] = next_hop_group_entry;
+            }
+            else {
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp]: Updated as suggested.

File: orchagent/vnetorch.cpp [hunk 7]
```diff
@@ -648,29 +831,126 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
...
+                {
+                    removeNextHopGroup(vnet, nexthops, vrf_obj);
+                }
+                return false;
+            }
         }
-    }
 
-    if (op == SET_COMMAND)
+        if (it_route != syncd_tunnel_routes_[vnet].end())
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp:887]: Added comments for this part.

File: tests/test_vnet.py (hunk 1 above)
Comment (shi-su) [reply tests/test_vnet.py:1327]: Added a test case as suggested.

File: tests/test_vnet.py [hunk 8]
```diff
@@ -790,7 +902,7 @@ def test_vnet_orch_1(self, dvs, testlog):
         vnet_obj.check_del_vnet_routes(dvs, 'Vnet_2001')
 
         delete_vnet_routes(dvs, "100.100.1.1/32", 'Vnet_2000')
-        vnet_obj.check_del_vnet_routes(dvs, 'Vnet_2001')
+        vnet_obj.check_del_vnet_routes(dvs, 'Vnet_2000')
```
Comment (prsunny) [inline tests/test_vnet.py:905]: Why is this changed?

Comment (shi-su) [reply tests/test_vnet.py:905]: I still believe this seems more like a typo, though I do not recall how I noticed this (maybe I added some more checks in the function and the test case failed). As we just removed a route from `Vnet_2000`, why do we want to check deleted route in `Vnet_2001`? I checked other calls of the same function and nowhere else use it in this way. I think it passed previously only because `check_del_vnet_routes` did not really check anything.

File: orchagent/vnetorch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.cpp]: Added support for multiple endpoints to share the same vni in the latest commit.

File: orchagent/vnetorch.h [hunk 9]
```diff
@@ -246,11 +247,12 @@ class VNetOrch : public Orch2
...
     {
-        { "endpoint",    REQ_T_IP },
-        { "ifname",      REQ_T_STRING },
-        { "nexthop",     REQ_T_STRING },
-        { "vni",         REQ_T_UINT },
-        { "mac_address", REQ_T_MAC_ADDRESS },
+        { "endpoint",               REQ_T_IP_LIST },
+        { "ifname",                 REQ_T_STRING },
+        { "nexthop",                REQ_T_STRING },
+        { "vni",                    REQ_T_STRING },
```
[the remark this answers is not in the document]
Comment (shi-su) [reply orchagent/vnetorch.h:253]: I do not think it would, since we will convert it to the right format. I think the only difference is to process the values in the request parser or vnetorch.

File: tests/test_vnet.py (hunk 8 above)
Comment (prsunny) [reply tests/test_vnet.py:905]: ok

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
@dgsudharsan , got to know that you were reviewing the PR. Please feel free to provide comments and we can address in a separate PR

**Comment by sachinv-msft** [comment]
@prsunny Cisco requested this PR for 202012 branch, could you review and recommend next steps/add the necessary labels?

**Comment by shi-su** [comment]
> @prsunny Cisco requested this PR for 202012 branch, could you review and recommend next steps/add the necessary labels?

https://github.com/Azure/sonic-swss/pull/2104 will take the changes to 202012 branch. Should be merging soon.
